### PR TITLE
RealmChangedListener now gets called on the same thread doing the commit

### PR DESCRIPTION
**Pull request by cmelchior**
This PR makes the OnRealmChanged callback behave more as expected and puts it in line with Realm-Cocoa behaviour.

Previously, callbacks registered on the thread doing the commit was not triggered, this is now the case. In order avoid weird cases like be below. The OnRealmChangedListener is called immediately instead of using the Handler mechanism.

```
// If using Handler messages, the following would be possible. By calling listeners at once, this is avoided.
realm.beginTransaction();
...
realm.commitTransaction();
realm.addChangeListener(new RealmChangedListener() {
  // This would get triggered even though it was added after the commit.
});
```

@emanuelez @kneth @jpsim @bmunkholm

### CODE REVIEW

File: realm/src/androidTest/java/io/realm/NotificationsTest.java [hunk 1]
```diff
@@ -304,4 +316,33 @@ public void testHandlerNotRemovedToSoon() {
...
+                success.set(true);
+            }
+        });
+        realm.beginTransaction();
+        realm.createObject(AllTypes.class);
+        realm.commitTransaction();
+        assertTrue(success.get());
+    }
+
+    public void testEmptyCommitTriggerOnChangeHandler() {
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/NotificationsTest.java]: Confusing test name. We don't use the word handler anywhere else, unless to specify the Android's Handler class

**Comment by emanuelez** [comment]
I'm not sure I like the calling of callbacks even if setup after a commit. It sounds counter-intuitive

**Comment by cmelchior** [comment]
Not sure what you mean?  You mean that triggering the listener immediately is wrong?

**Comment by emanuelez** [comment]
No, I'm referring to what you describe in the code snippet of your first comment.

**Comment by kneth** [comment]
Just thinking about timing/ordering of operations: will this change imply that the change listener will be called when the Realm is initially opened (`Realm.initializeRealm()` might do a commit).

**Comment by cmelchior** [comment]
It will, which might be a bit odd (although technically correct). I checked Realm-Cocoa and as far as I could tell they do the same.

**Comment by tgoyne** [comment]
We had to do some non-trivial refactoring to eliminate spurious notifications from files being opened. The actual schema initialization happens before the user gets the chance to add a listener, and when Realms are opened with no initialization/migration needed we're careful to avoid committing a write transaction.

**Comment by tgoyne** [comment]
Also, for the multi-process case we have to avoid even starting a write transaction when schema changes aren't needed, so that extensions don't hang on load when the main app is suspended during a write transaction (writing in that scenario obviously still doesn't work).

**Comment by cmelchior** [comment]
Good points @tgoyne. We already grab a transaction and abort it if the schema is detected to be up to date. By executing the listeners immediately on the same thread we also go clear of the issue of getting RealmChange notifications for just opening (and creating) the initial Realm.

We have not implemented multiprocess support just yet, so any issues in this area will have to be fixed when that is introduced. I especially suspect that we need to begin a write transaction earlier to avoid  race conditions when creating the initial schema.

**Comment by emanuelez** [comment]
:+1:

**Comment by cmelchior** [comment]
Bump @kneth

**Comment by kneth** [comment]
:+1:

**Comment by beeender** [comment]
Hi

I still prefer the previous implementation.
Now, it would trigger ConcurrentModificationException if user is not careful enough, and actually some very common user cases  would trigger ConcurrentModificationException.

eg.
sendNotifications triggered the pageAdapter to refresh in onChange. And pageAdapter created some fragment which also has its realm instance and listener since the fragment wants to keep the instance in its lifecycle. Then when the addListener called in the fragment, ConcurrentModificationException happens.

Yes, post the fragment running would solve the problem, but sometimes this is not easy to reproduce & debug.

**Pull request merged**